modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageCompactionTrigger.java

```diff
@@ -0,0 +1,270 @@
+            String localNodeName,
+            KeyValueStorage storage,
+            MetaStorageManagerImpl metaStorageManager,
+            ScheduledExecutorService scheduler
+    ) {
+        this.localNodeName = localNodeName;
+        this.storage = storage;
+        this.metaStorageManager = metaStorageManager;
+        this.scheduler = scheduler;
+
+        storage.registerCompactionListener(
+                compactionRevision -> inBusyLockSafe(busyLock, () -> onCompactionCompleteLocallyBusy(compactionRevision))
+        );
+
+        metaStorageManager.addElectionListener(newLeader -> inBusyLockSafe(busyLock, () -> onLeaderElectedBusy(newLeader)));
+    }
+
+    @Override
+    public CompletableFuture<Void> startAsync(ComponentContext componentContext) {
+        return nullCompletedFuture();
+    }
+
+    @Override
+    public CompletableFuture<Void> stopAsync(ComponentContext componentContext) {
+        if (!stopGuard.compareAndSet(false, true)) {
+            return nullCompletedFuture();
+        }
+
+        busyLock.block();
+
+        cancelLastScheduledFutureBusy();
+
+        return nullCompletedFuture();
+    }
+
+    private void doCompactionBusy() {
+        lock.lock();
+
+        try {
+            if (!isLocalNodeLeader) {
+                return;
+            }
+
+            HybridTimestamp candidateCompactionRevisionTimestamp = createCandidateCompactionRevisionTimestampBusy();
+
+            Long newCompactionRevision = calculateCandidateCompactionRevisionBusy(candidateCompactionRevisionTimestamp);
+
+            if (newCompactionRevision == null) {
+                scheduleNextCompactionBusy();
+            } else {
+                metaStorageManager
+                        .sendCompactionCommand(newCompactionRevision)
+                        .thenCompose(
+                                unused -> inBusyLockAsync(busyLock, () -> awaitCompleteCompactionLocallyFutureBusy(newCompactionRevision))
+                        )
+                        .thenRun(() -> inBusyLock(busyLock, this::scheduleNextCompactionBusy))
+                        .whenComplete((unused, throwable) -> {
+                            if (throwable != null) {
+                                Throwable cause = unwrapCause(throwable);
+
+                                if (!(cause instanceof NodeStoppingException)) {
+                                    LOG.error(
+                                            "Unknown error on new metastorage compaction revision: ",
+                                            cause,
+                                            newCompactionRevision
+                                    );
+
+                                    inBusyLockSafe(busyLock, this::scheduleNextCompactionBusy);
+                                }
+                            }
+                        });
+            }
+        } catch (Throwable t) {
+            LOG.error("Unknown error on new metastorage compaction revision", t);
```

Comment:
```suggestion
            LOG.error("Unknown error on new metastorage compaction revision scheduling", t);
```

Something like that, to make it easier to differentiate problems while compacting and in this try block